metaflow/client/core.py

```diff
@@ -1123,6 +1124,202 @@ def _iter_filter(self, x):
+                field_name = "foreach-indices"
+                field_value = self.metadata_dict.get("foreach-indices-truncated")
+        return field_name, field_value
+
+    def _get_related_tasks(self, relation_type: str) -> Dict[str, List[str]]:
+        start_time = time.time()
+        flow_id, run_id, _, _ = self.path_components
+        steps = (
+            self.metadata_dict.get("previous_steps")
+            if relation_type == "ancestor"
+            else self.metadata_dict.get("successor_steps")
+        )
+
+        if not steps:
+            return {}
+
+        field_name, field_value = self._get_metadata_query_vals(
+            flow_id,
+            run_id,
+            len(self.metadata_dict.get("foreach-indices", [])),
+            steps,
+            relation_type,
+        )
+
+        cur_time = time.time()
+
+        return {
+            step: [
+                f"{flow_id}/{run_id}/{step}/{task_id}"
+                for task_id in self._metaflow.metadata.filter_tasks_by_metadata(
+                    flow_id, run_id, step, field_name, field_value
+                )
+            ]
+            for step in steps
+        }
+
+    @property
+    def immediate_ancestors(self) -> Dict[str, List[str]]:
+        """
+        Returns a dictionary of immediate ancestors task pathspecs of this task for each
+        previous step.
+
+        Returns
+        -------
+        Dict[str, List[str]]
+            Dictionary of immediate ancestors of this task. The keys are the
+            names of the ancestor steps and the values are the corresponding
+            task pathspecs of the ancestors.
+        """
+        return self._get_related_tasks("ancestor")
+
+    @property
+    def immediate_successors(self) -> Dict[str, List[str]]:
+        """
+        Returns a dictionary of immediate successors task pathspecs of this task for each
+        previous step.
+
+        Returns
+        -------
+        Dict[str, List[str]]
+            Dictionary of immediate successors of this task. The keys are the
+            names of the successor steps and the values are the corresponding
+            task pathspecs of the successors.
+        """
+        return self._get_related_tasks("successor")
+
+    @property
+    def immediate_siblings(self) -> Dict[str, List[str]]:
+        """
+        Returns a dictionary of closest sibling task pathspecs of this task for each
+        sibling step.
+
+        Returns
+        -------
+        Dict[str, List[str]]
+            Dictionary of closest siblings of this task. The keys are the
+            names of the current step and the values are the corresponding
+            task pathspecs of the siblings.
+        """
+        flow_id, run_id, step_name, _ = self.path_components
+
+        foreach_stack = self.metadata_dict.get("foreach-stack", [])
+        foreach_step_names = self.metadata_dict.get("foreach-step-names", [])
+        if len(foreach_stack) == 0:
```

Comment:
For a user exposed function called immediate_siblings, we should handle the static case too ideally (or just change the name to make it clear it's just for foreach). For example, if I have A -> B/C/D -> E, I would expect that in B, I could get the siblings C and D. Similarly, even if I am not a new foreach split, something like A -> foreach B -> C/D/E -> BJoin -> AJoin, from C, I should get D and E (the right ones). I think this function is excluding those right?